### PR TITLE
Add support for recording feature values directly via a new TPL builtin, execCtxRecordFeature.

### PR DESCRIPTION
**Pull request by lmwnshn**
This PR adds support for directly recording to the metrics feature vector via a new TPL builtin `@execCtxRecordFeature`.

Summary:

- Counters are implemented to follow the definition in [brain_defs.h](https://github.com/cmu-db/terrier/pull/1134/files#diff-4c774f0c39f07eb16ef19af8143d421aR8), courtesy of @17zhangw 
- All recording is gated by [CompilationContext::IsCountersEnabled()](https://github.com/cmu-db/terrier/pull/1134/files#diff-a8dce8335eae87fbb935cabb7632ef0cR121), which is currently not exposed but in theory could be hooked up to the settings manager. All the code in this PR should be easily ripped out in the future once we have real table statistics.
- Counters are stashed into a TPL program's `queryState` as the counters may need to be maintained across TPL functions within the same query.
- Convenience functions are defined in `OperatorTranslator`: `CounterDeclare`, `CounterSet`, `CounterAdd`. These may be useful for adding new counters. They automatically use the `if (IsCountersEnabled()) ...` guard.
- Modified the [BwTree](https://github.com/cmu-db/terrier/pull/1134/files#diff-f0c28d9553fd33b2315722bd96f3c6cfR6794) to expose a `GetSize()` method, as well as our [hash index](https://github.com/cmu-db/terrier/pull/1134/files#diff-15f0a2eea833a260da9002625361dff7R74). This is for `IDX_SCAN::NUM_ROWS`.

Known issues:

- All counters force serial pipelines right now. This will be addressed as me and @17zhangw sync up...

### CODE REVIEW

File: src/execution/compiler/operator/seq_scan_translator.cpp
Comment (17zhangw) [inline src/execution/compiler/operator/seq_scan_translator.cpp:221]: This is a small concern. `num_scans_` records the number of tuples that match the filter and are passed on to the next translator. This means that as far as the model is concerned, the cost of scanning 1 million tuples to emit 1 tuple is the same as the cost of scanning 1 tuple to emit 1 tuple.

Is there a way to get the raw number of tuples in the table (or at least a close approximation)?

File: src/execution/compiler/operator/sort_translator.cpp
Comment (17zhangw) [inline src/execution/compiler/operator/sort_translator.cpp:258]: Since we don't necessarily want to track data uniqueness, this might be the best we can do? (although the prediction results will be slightly affected since `sort_build` does depend on the input data cardinality).

@malin1993ml

File: src/include/brain/brain_defs.h
Comment (17zhangw) [inline src/include/brain/brain_defs.h:98]: In response to the TODO, it is set to 0 when using the NoOpResultConsumer

File: src/include/brain/brain_defs.h
Comment (17zhangw) [inline src/include/brain/brain_defs.h:16]: Is this TODO still current?

File: src/include/brain/brain_defs.h
Comment (17zhangw) [inline src/include/brain/brain_defs.h:59]: In response to TODO, cardinality can probably be set to num_rows without anything going noticeably wrong (since a seq scan doesn’t really depend on table cardinality).

On a separate note, should probably amend the num_rows definition once a previous comment gets resolved?

**Review by 17zhangw — CHANGES_REQUESTED** [review]
Generally LGTM. Just a few questions.

File: src/execution/compiler/operator/seq_scan_translator.cpp
Comment (malin1993ml) [reply src/execution/compiler/operator/seq_scan_translator.cpp:221]: I think the concern is valid. Maybe directly utilize the `SqlTable::GetNumTuple` function?

File: src/execution/compiler/operator/sort_translator.cpp
Comment (malin1993ml) [reply src/execution/compiler/operator/sort_translator.cpp:258]: Yeah I think this is the best we can do. I would add a TODO here to document that in theory we should use the number of unique tuples as cardinality, but given the potential overhead and engineering complexity we just use the size of the sorter.

File: src/include/brain/brain_defs.h
Comment (malin1993ml) [inline src/include/brain/brain_defs.h:41]: Do we still need this if the feature doesn't actually exist? Also, it seems that we haven't modified the `index_join_translator` to add the counters, especially for the `num_loops` feature?

File: src/include/brain/brain_defs.h
Comment (malin1993ml) [inline src/include/brain/brain_defs.h:72]: This is just the same as `num_rows`.

File: src/include/brain/brain_defs.h
Comment (malin1993ml) [inline src/include/brain/brain_defs.h:78]: Same here.

File: src/include/brain/brain_defs.h
Comment (malin1993ml) [inline src/include/brain/brain_defs.h:84]: Same here.

File: src/include/brain/brain_defs.h
Comment (malin1993ml) [inline src/include/brain/brain_defs.h:131]: Since none of these features actually exist, shouldn't we just remove them?

File: src/include/execution/compiler/compilation_context.h
Comment (malin1993ml) [inline src/include/execution/compiler/compilation_context.h:121]: Yeah I think we should probably set this to `false` by default.

File: src/execution/compiler/operator/sort_translator.cpp
Comment (malin1993ml) [reply src/execution/compiler/operator/sort_translator.cpp:258]: I was also running TPCC just now with this branch. I noticed that this value as well as the cardinality value for `SORT_ITERATE` are recorded as a very high corrupted value. This probably needs double-check.

**Comment by malin1993ml** [comment]
While I was testing against the TPCH queries, I noticed that there's another high-level thing: there're certain translators that have [arithmetic features](https://github.com/cmu-db/terrier/blob/master/src/include/brain/brain_defs.h#L40), such as the predicate for seq scan, index scan, or joins. Those features are already recorded in the pipeline features, but we need to update their counters as well. Thus, we need another API in the `ExecutionContext`:
`RecordArithmeticFeature(pipeline_id_t pipeline_id, translator_id_t translator_id, uint32_t value);`
This function will **multiply** the `num_rows`  and `cardinality` for all the arithmetic features by `value`. `num_rows` and `cardinality` are equal because arithmetics don't really care the uniqueness. We need to add this to almost all the translators you have. It's multiplication because we may have merged multiple arithmetics under the same type together (the original values would just be the number of occurrences). For most translators, this value would just be equal to the `num_rows` value of the primary feature of the translator (such as `num_rows` for the `SEQ_SCAN` OU of the `seq_scan_translator`). I think the only two exceptions would be `HASHJOIN_PROBE` and `IDX_SCAN`, where we're going to use their `cardinality` to set the values for the arithmetics in the same translator. Let me know whether this makes sense, and we can also chat offline. The rest of the PR is awesome.

File: src/execution/compiler/operator/seq_scan_translator.cpp
Comment (lmwnshn) [reply src/execution/compiler/operator/seq_scan_translator.cpp]: Changed it to add the tuple count from the VPI.

For the record, looks like `SqlTable::GetNumTuple` is borked for getting approximate counts.

File: src/execution/compiler/operator/sort_translator.cpp
Comment (lmwnshn) [reply src/execution/compiler/operator/sort_translator.cpp:258]: It works for small test cases for me.. Does it go above int32_max? I wonder if we are hitting #1049.

File: src/execution/compiler/operator/seq_scan_translator.cpp
Comment (malin1993ml) [reply src/execution/compiler/operator/seq_scan_translator.cpp]: I think your change works.

File: src/execution/compiler/operator/sort_translator.cpp
Comment (malin1993ml) [reply src/execution/compiler/operator/sort_translator.cpp:258]: I don't think it goes above int32_max. I just directly compiled your branch and executed TPCC against it with `--metrics_pipeline` turned on. Maybe try TPCC locally to see whether you can reproduce?

Comment (lmwnshn) [reply src/execution/compiler/operator/sort_translator.cpp:258]: Fixed, it was a bug with bytecode generation.

File: src/include/brain/brain_defs.h
Comment (lmwnshn) [reply src/include/brain/brain_defs.h:98]: I think in practice I can always just record the feature as 1, then, right? I'm not sure I know which consumer is consuming it.

File: src/include/brain/brain_defs.h
Comment (lmwnshn) [reply src/include/brain/brain_defs.h:16]: Yes; an aggregation typically smashes a whole bunch of values into a single value right? When will there be more than 1 value in an aggregation's output? I cannot think of any right now.

File: src/include/brain/brain_defs.h
Comment (lmwnshn) [reply src/include/brain/brain_defs.h]: Done.

File: src/include/brain/brain_defs.h
Comment (malin1993ml) [reply src/include/brain/brain_defs.h]: I think you can leave this particular field as it is. In practice it will be set outside, and you're right that you won't know which consumer it is.

File: src/include/brain/brain_defs.h
Comment (malin1993ml) [reply src/include/brain/brain_defs.h:16]: Well in the case of a group by there will be more values (same number of rows as the distinct values in the group by column).

File: src/include/brain/brain_defs.h
Comment (lmwnshn) [reply src/include/brain/brain_defs.h]: Removed.

File: src/include/brain/brain_defs.h
Comment (lmwnshn) [reply src/include/brain/brain_defs.h]: Removed.

File: src/include/brain/brain_defs.h
Comment (lmwnshn) [inline src/include/brain/brain_defs.h:67]: @malin1993ml @17zhangw Can you confirm this N/A part? Or should cardinality be the same as NUM_ROWS?

Comment (malin1993ml) [reply src/include/brain/brain_defs.h:67]: `num_rows` and `cardinality` would actually be the same as a regular index scan. That is, the size of the index and the size of the scan per `idx_scan` invocation. If calculating a per-invocation value is not convenient, you can also just use the value of the last invocation as an approximation.

Comment (lmwnshn) [reply src/include/brain/brain_defs.h:67]: I do not understand what num_loops would be recorded as then (the number of times the outer loop is accessed).

Comment (lmwnshn) [reply src/include/brain/brain_defs.h]: That said, I have pushed my changes.

File: src/include/brain/brain_defs.h
Comment (lmwnshn) [reply src/include/brain/brain_defs.h:23]: Got it.

**Comment by lmwnshn** [comment]
I think I have added the arithmetic features and addressed review comments, could you take a look to see if I understood what is needed? Thanks! @malin1993ml @17zhangw

File: src/include/brain/brain_defs.h
Comment (17zhangw) [reply src/include/brain/brain_defs.h]: I took a look at the changes to the `IndexJoinTranslator`. So the existing `num_scans_index_` essentially tracks how many tuples are retrieved from the index. So this would be recorded as the `cardinality` (which is what is done now).

For `num_loops`, it would essentially be tracking the number of times `@indexIteratorScanKey(&index_iter)` is invoked (which equals # tuples emit by the index join plan's child and equals # times the index is looked up on).

File: src/execution/compiler/operator/operator_translator.cpp
Comment (malin1993ml) [reply src/execution/compiler/operator/operator_translator.cpp:185]: Thanks for following up on this, William! I think you're correct that we need to differentiate the translator ids when doing the arithmetic multiplication. And we do need to modify the `operating_unit_recorder` to only merge arithmetics of the same translator.

File: src/brain/operating_unit_recorder.cpp
Comment (lmwnshn) [reply src/brain/operating_unit_recorder.cpp:388]: Fixed with your suggestion of moving things [around](https://github.com/cmu-db/terrier/pull/1134/commits/073ab53e009633fee373f1fe185dc8043f2366ff#diff-74b764e1da242750e292732b021c45e2R432).

File: src/brain/operating_unit_recorder.cpp
Comment (lmwnshn) [inline src/brain/operating_unit_recorder.cpp:225]: @17zhangw

File: src/execution/compiler/operator/operator_translator.cpp
Comment (lmwnshn) [reply src/execution/compiler/operator/operator_translator.cpp:198]: Modified by having each translator denote whether it should own its features or pass it through to some other translator. Right now, projection and nested_loop_join passes it through.

File: src/execution/compiler/operator/index_join_translator.cpp
Comment (malin1993ml) [inline src/execution/compiler/operator/index_join_translator.cpp:116]: To match the models, can we change this to per-loop num scans, if num loops > 0 (i.e., int(num_scans_index_ / num_loops_))?

File: src/brain/operating_unit_recorder.cpp
Comment (malin1993ml) [inline src/brain/operating_unit_recorder.cpp:222]: We also need to check that the pipeline ids are the same before we merge features here, since now we differentiate the multiplication by the pipeline ids.

**Review by malin1993ml — CHANGES_REQUESTED** [review]
I think overall this is great! I've checked the TPCC recordings, except some minor comments I left below, all the other features look correct!

I also ran the `tpch_runner` to check the counters as well, but I observed something weird. Q1-p3 has `num_rows` > 0 for `AGG_BUILD`, but `cardinality` (size of the hash table) = 0. And subsequently, Q1-p2 and Q1-p3 have `num_rows` = 0 for all their OUs. This is wrong because an aggregation with `num_rows` > 0 should always have return values. I don't know whether this is related to the counters we added, but could you double-check on this?

Q18-p6 also has all the OUs with `num_rows` = 0, which also seems suspicious to me.

And I also noticed some arithmetic features for TPCH queries still have `num_rows` = 0 despite all the translators in the pipeline have `num_rows` > 0. The last `OP_INTEGER_PLUS_OR_MINUS` feature for Q4-p3 is an example of this. This is less important, but could you double-check as well if you have time?

**Comment by lmwnshn** [comment]
@thepinetree This is the branch, you can probably just build on top of it. We might want to call briefly.

File: src/execution/compiler/operator/output_translator.cpp
Comment (malin1993ml) [inline src/execution/compiler/operator/output_translator.cpp:58]: A minor thing: we should remove setting the cardinality for `OUTPUT` since you don't really know which consumer it is and it would be set outside.

File: src/execution/compiler/operator/hash_aggregation_translator.cpp
Comment (lmwnshn) [inline src/execution/compiler/operator/hash_aggregation_translator.cpp:390]: I am trying to convince myself that this is correct.

File: src/brain/operating_unit_recorder.cpp
Comment (lmwnshn) [reply src/brain/operating_unit_recorder.cpp:234]: Added a check on the set of translators.

File: src/execution/compiler/operator/static_aggregation_translator.cpp
Comment (lmwnshn) [inline src/execution/compiler/operator/static_aggregation_translator.cpp:188]: And here too.

File: src/execution/compiler/operator/hash_aggregation_translator.cpp
Comment (malin1993ml) [reply src/execution/compiler/operator/hash_aggregation_translator.cpp:390]: Emm I'm surprised that we even need to do this. I think in theory we just need to multiply the arithmetic features by `num_agg_inputs_` (which you already did). By default each OU has a [`num_rows` = 1](num_agg_inputs_), and the arithmetic OUs will be invoked by `num_agg_inputs_` times. Adding another counter does not seem to be the best direction for me. Unless the arithmetics have a `num_rows` = 0 to begin with? Then that might be a bug in the operating unit recorder. Also, the multiplication of `num_agg_outputs_` in the `AGG_ITERATE` phase seems redundant to me...

Comment (lmwnshn) [reply src/execution/compiler/operator/hash_aggregation_translator.cpp]: Removed, it was a different bug.

Could you elaborate on the redundancy?

**Comment by lmwnshn** [comment]
@malin1993ml I had a bug where I was multiplying the arithmetic OU every iteration instead of at the end. Due to #1049, this bug overwrites random memory and may explain weird things that you've seen before.

SF1, I am not seeing anything too weird for e.g. Q1 now, I think:

```
0, 3, 0, 6, OP_INTEGER_COMPARE;AGG_BUILD;OP_INTEGER_PLUS_OR_MINUS;OP_DECIMAL_MULTIPLY;SEQ_SCAN;OP_DECIMAL_PLUS_OR_MINUS, 6001215;5914748;5914748;17744244;6001215;59147480, 4;128;4;8;4;8, 1;32;1;1;1;1, 6001215;4;5914748;17744244;6001215;59147480, 1;0.343284;1;1;1;1, 0;0;0;0;0;0, 1599493930515711, 0, 0, 0, 0, 0, 0, 0, 0, 47512, 2215103
0, 2, 0, 2, AGG_ITERATE;SORT_BUILD, 4;4, 188;128, 40;32, 4;4, 1;0.34375, 0;0, 1599493932730821, 0, 0, 0, 0, 0, 0, 0, 0, 45248, 6
0, 1, 0, 2, SORT_ITERATE;OUTPUT, 4;4, 188;188, 40;40, 4;1, 1;1, 0;0, 1599493932730828, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2
```

File: src/execution/compiler/operator/hash_aggregation_translator.cpp
Comment (malin1993ml) [reply src/execution/compiler/operator/hash_aggregation_translator.cpp]: I see! In that case, wouldn't the `num_rows` and `cardinality` features for `AGG_BUILD` and `AGG_ITERATE` also be recorded many times instead of just one time in the end? (Though the eventual value is correct.) Maybe worth double-checking other counters.

For the redundancy - Basically I'm wondering would the `AGG_BUILD` and `AGG_ITERATE` have the same translator id or not? If they have the same translator id, then all the arithmetics would be multiplied twice in both of the two pipelines, right?

File: src/execution/compiler/operator/delete_translator.cpp
Comment (lmwnshn) [inline src/execution/compiler/operator/delete_translator.cpp:49]: Move this to FinishPipelineWork.

File: src/execution/compiler/operator/output_translator.cpp
Comment (lmwnshn) [inline src/execution/compiler/operator/output_translator.cpp:27]: Move to InitializeQueryState.

File: src/execution/compiler/operator/seq_scan_translator.cpp
Comment (lmwnshn) [inline src/execution/compiler/operator/seq_scan_translator.cpp:302]: Move this to FinishPipelineWork.

File: src/execution/compiler/operator/update_translator.cpp
Comment (lmwnshn) [inline src/execution/compiler/operator/update_translator.cpp:86]: Move this to FinishPipelineWork.

**Review by lmwnshn — COMMENTED** [review]
Fix me.

**Review by malin1993ml — CHANGES_REQUESTED** [review]
I think this is good! A few minor things and then I think we're done!

File: src/execution/compiler/operator/hash_aggregation_translator.cpp
Comment (malin1993ml) [inline src/execution/compiler/operator/hash_aggregation_translator.cpp:479]: Let's add the multiplication back here...

**Review by malin1993ml — CHANGES_REQUESTED** [review]
I actually took my last review back... I just realized that the arithmetics on two sides of an operator would end up in two different pipelines. And since we record the features at a per-pipeline basis, the two multiplication actually should not affect each other at all, unless you put the two sides in a single pipeline (which I think won't happen in our engine). So we don't need to remove those multiplications and instead should just add the multiplication back to the `AGG_ITERATE` side. Please think about whether this makes sense.

File: src/execution/compiler/operator/hash_aggregation_translator.cpp
Comment (lmwnshn) [reply src/execution/compiler/operator/hash_aggregation_translator.cpp:479]: Added it to FinishPipelineWork (the convention I'm running with now is for all recording to happen there by default).

Your reasoning makes sense, the translator ID is the same but the pipeline ID is not the same.

**Review by malin1993ml — APPROVED** [review]

**Comment by malin1993ml** [comment]
Awesome! I think there're still tests failing? After the build passes I think we're ready to merge!

**Comment by lmwnshn** [comment]
Interesting, I don't think I touched the index create test and I can't reproduce this locally. Going to merge master and see what happens.

File: src/execution/sema/sema_checking.cpp
Comment (lmwnshn) [inline src/execution/sema/sema_checking.cpp:229]: Hey @pmenon, could you quickly weigh in on this change?

left: MemberExpr, it is queryState.some_uint32
right: LitExpr, integer literal in particular, treated as int32

I'm not sure if:
- there is a cast for narrowing/widening integers
- there is a more fine-grained way of comparing types here than IsIntegerType which should be used

Thanks!

Comment (pmenon) [reply src/execution/sema/sema_checking.cpp:229]: There isn't a means to down- or up-cast primitive integers at the moment. For now, I think what you have is okay. But, put a `TODO` putting the blame on me and revisiting with a fix. Also, create an issue for it.

Comment (lmwnshn) [reply src/execution/sema/sema_checking.cpp:230]: #1174. Thanks!

**Review by malin1993ml — APPROVED** [review]
This is great! I've tested that this works on both TPCC and TPCH. Thanks for the awesome work, Wan!

Only thing is that there still seems to be a unit test failing. I think we're good to go once the tests pass.

File: src/execution/sema/sema_checking.cpp
Comment (lmwnshn) [inline src/execution/sema/sema_checking.cpp]: whoops

**Pull request merged**
